pkg/controller/deployment/sync.go

```diff
@@ -330,7 +330,8 @@ func (dc *DeploymentController) getNewReplicaSet(d *extensions.Deployment, rsLis
 		}
 		// Matching ReplicaSet is not equal - increment the collisionCount in the DeploymentStatus
 		// and requeue the Deployment.
-		if !deploymentutil.EqualIgnoreHash(&d.Spec.Template, &rs.Spec.Template) {
+		if !deploymentutil.EqualIgnoreHash(&d.Spec.Template, &rs.Spec.Template) ||
+			!reflect.DeepEqual(&d.Labels, &rs.Labels) {
```

Comment:
A change in labels happened to trigger the problem in our case, but the root problem is that we should trigger collision avoidance if the name collides with a ReplicaSet that we don't own (in the sense of ControllerRef).

I think the right fix is to check whether the Deployment UID matches the one returned by `metav1.GetControllerOf()` (if any).

Reply:
To reiterate, we should trigger the collision avoidance when:
1. PodTemplateSpec of the replicaset is not semantically deep equal with that of the deployment
2. the replicaset's controller is the deployment

```
controllerRef := metav1.GetControllerOf(rs)
if !deploymentutil.EqualIgnoreHash(&d.Spec.Template, &rs.Spec.Template) ||
    (controllerRef != nil && reflect.DeepEqual(d.UID, controllerRef.UID)) {
```

Reply:
Do you mind to explain more about this? Without explicitly checking for `.Labels` change, the integration test cannot be passed.

Reply:
Not quite. We also want to trigger collision avoidance if the RS is an orphan, because it means we don't own it.

I think it would be easier to read if we invert the condition. There is only one specific case where we want to skip collision avoidance (note you don't need `DeepEqual` for `UID`):

```go
// If we own this RS, and its PodTemplateSpec matches what we want, it's our new RS.
if controllerRef != nil && controllerRef.UID == d.UID && deploymentutil.EqualIgnoreHash(&d.Spec.Template, &rs.Spec.Template) {
  createdRs = rs
  err = nil
  break
}
// collision avoidance
...
```

Reply:
Yes, I missed out the case when the RS is an orphan, in which we want to trigger the collision avoidance too. Inverting the conditions is clearer - I will add more comments to clarify it.

Reply:
To sum up, the collision avoidance mechanism is triggered when:
1. the replicaset is an orphan: the replicaset was ours, but not anymore; slow cache or fast resync may result in fetching the orphaned replicaset
2. the replicaset is owned by another object: it is not our replicaset and we do not have any new replicaset now, so we should trigger the mechanism to create one
3. the replicaset has different PodTemplateSpec: in other words, the replicaset has not yet picked up the deployment's PodTemplateSpec change (and thus not orphaned yet)